controller/server.go

```diff
@@ -0,0 +1,281 @@
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
+
+	"github.com/elastic/elastic-agent-shipper/config"
+	"github.com/elastic/elastic-agent-shipper/monitoring"
+	"github.com/elastic/elastic-agent-shipper/output"
+	"github.com/elastic/elastic-agent-shipper/output/elasticsearch"
+	"github.com/elastic/elastic-agent-shipper/output/kafka"
+	"github.com/elastic/elastic-agent-shipper/output/logstash"
+	"github.com/elastic/elastic-agent-shipper/queue"
+	"github.com/elastic/elastic-agent-shipper/server"
+
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// Output describes a typical output implementation used for running the server.
+type Output interface {
+	Start() error
+	Wait()
+}
+
+// ServerRunner starts and gracefully stops the shipper server on demand.
+// The server runner is not re-usable and should be abandoned after `Close` or canceled context.
+// under V2, this object will assume control of any unit state changes that happen internally,
+// and will be responsible for marking the shipper's units as failed, started, etc
+// This is because the input and output unit will map to different internal components; queue/output for the Output unit, gRPC for tine Input unit.
+type ServerRunner struct {
+	log *logp.Logger
+	cfg config.ShipperRootConfig
+	// to protect the `Close` function from multiple concurrent calls,
+	// so we avoid race conditions
+	closeOnce sync.Once
+	// to make sure that `Close` is not called before `Start` fully finished
+	startMutex sync.Mutex
+
+	server     *grpc.Server
+	shipper    server.ShipperServer
+	queue      *queue.Queue
+	monitoring *monitoring.QueueMonitor
+	out        Output
+
+	// units for reporting state
+	// In the future, we'll have a dedicated input unit for GRPC
+	// Right now, the mapping is:
+	// output unit -> queue/output state
+	// input unit -> gRPC state
+	outUnit *client.Unit
+	inUnit  *client.Unit
+}
+
+// NewOutputServer initializes the shipper queue and output based on the config received
+// errors and state will be sent upstream based on if the error was in an input or output subsystem.
+func NewOutputServer(cfg config.ShipperRootConfig, grpcTLS credentials.TransportCredentials, outUnit *client.Unit, inUnit *client.Unit) (r *ServerRunner, err error) {
+	r = &ServerRunner{
+		log:     logp.L(),
+		cfg:     cfg,
+		outUnit: outUnit,
+		inUnit:  inUnit,
+	}
+
+	r.reportState(r.outUnit, "initializing the queue...", client.UnitStateStarting)
+	r.queue, err = queue.New(cfg.Shipper.Queue)
+	if err != nil {
+		msg := fmt.Errorf("couldn't create queue: %w", err)
+		r.reportState(r.outUnit, msg.Error(), client.UnitStateFailed)
+		return nil, msg
+	}
+
+	r.reportState(r.outUnit, "queue was initialized.", client.UnitStateStarting)
```

Comment:
The underlying `UpdateState` call to the agent will perform a Checkin with the agent, and the agent will likely include this message in its own logs (it definitely will for state changes, I have to test to confirm it also does this when the state doesn't update).

If the intent is to use this for debug logging you are putting a lot of traffic on the gRPC stream to the agent here.

I would recommend making a clear separation between logging and when you want to update your state with the agent. This pattern seems to be encouraging very rapid checkins with the agent.

Reply:
Yah, because unit state tends to exist on a spectrum with `Stopping`, `Starting`, `Stopped`, `Healthy`, etc, it's never 100% clear when something should just be a log event, a state update, or both.

Reply:
My thinking was that it would be best to "centralize" the logs and treat (almost) everything like a status update, but now I'm not so sure.

Reply:
I would only send actual status changes to the agent, that is the intent. There is a bunch of machinery behind the status updates to calculate if the expected state of the unit is still the same and send it back. Each status update is effectively a checkin.

Most of these updates should probably just be log statements.

Reply:
Yah, agreed.